addons/microsoft_calendar/models/calendar.py

```diff
@@ -347,6 +347,23 @@ def _microsoft_values(self, fields_to_sync, initial_values={}):
 
         return values
 
+    def _ensure_attendees_have_email(self):
+        if self.env.context.get("_attendees_have_email", False):
+            return
+        invalid_events = ['\t- %s: %s' % (event.display_time, event.display_name)
+                          for event in self if any(not a.email for a in event.attendee_ids)]
+        if invalid_events:
```

Comment:
```
invalid_event_ids = self.filtered(lambda e: e.attendee_ids and any(not a.email for a in e.attendee_ids))
if invalid_event_ids:
    invalid_events =  ['\t- %s: %s' % (event.display_time, event.display_name)
                          for event in invalid_event_ids]
    invalid_events.sort()
    etc etc
```
That way, you avoid to read display_time, display_name of all events if it's not necessary. You only read it for the records that need it. You avoid a lot of computation.

Reply:
You could also optimize to avoid querying multiple calendar.attendee with the same partner. You know that all the corresponding events will trigger an error, you only need to raise the error. It may be not useful to display all the events information. What you want is that the partner is fixed once and for all.

Reply:
@arj-odoo I understand the logic, but I'm not sure it would be great from a UX perspective. Suppose the user has two incorrect events A and B, because they both have Foo as attendee and he has not email address. In this case, Odoo displays the validation error for A, but the user might ask "But why isn't there a problem with B, then ?"

I updated the code. Now it searches in `calendar.attendee`, as you suggested. I also "cut" the list before calling `.read(['display_time', 'display_name'])` to avoid computing values that will not be displayed.